Test failures with cirq==0.12

With the release of Cirq v0.12, several qsimcirq tests are now failing with error messages similar to the one below.

<details>

```
  ______________________________ test_empty_circuit ______________________________
  
      def test_empty_circuit():
  >       result = qsimcirq.QSimSimulator().simulate(cirq.Circuit())
  
  /project/qsimcirq_tests/qsimcirq_test.py:37: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  /tmp/tmp.1VNqjXUuKR/venv/lib/python3.6/site-packages/cirq/sim/simulator.py:429: in simulate
      program, study.ParamResolver(param_resolver), qubit_order, initial_state
  /tmp/tmp.1VNqjXUuKR/venv/lib/python3.6/site-packages/qsimcirq/qsim_simulator.py:427: in simulate_sweep
      params=prs, measurements={}, final_simulator_state=final_state
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <[AttributeError("'QSimSimulatorTrialResult' object has no attribute 'params'",) raised in repr()] QSimSimulatorTrialResult object at 0x7f791b4cae48>
  params = cirq.ParamResolver({}), measurements = {}
  final_simulator_state = cirq.StateVectorSimulatorState(state_vector=np.array([1.+0.j], dtype=complex64), qubit_map={})
  
      def __init__(
          self,
          params: study.ParamResolver,
          measurements: Dict[str, np.ndarray],
          final_simulator_state: QSimSimulatorState,
      ):
          super().__init__(
              params=params,
              measurements=measurements,
  >           final_simulator_state=final_simulator_state,
          )
  E       TypeError: __init__() got an unexpected keyword argument 'final_simulator_state'
  
  /tmp/tmp.1VNqjXUuKR/venv/lib/python3.6/site-packages/qsimcirq/qsim_simulator.py:56: TypeError
```

</details>

We need to bring qsim up-to-date with modern Cirq.

--- a/qsimcirq/qsim_simulator.py
+++ b/qsimcirq/qsim_simulator.py
@@ -81,21 +81,21 @@ class QSimSimulatorTrialResult(sim.StateVectorMixin, sim.SimulationTrialResult):
             state_vector = self.dirac_notation(3)
         else:
             state_vector = str(final)
-        return f'measurements: {samples}\noutput vector: {state_vector}'
+        return f"measurements: {samples}\noutput vector: {state_vector}"
 
     def _repr_pretty_(self, p: Any, cycle: bool) -> None:
         """Text output in Jupyter."""
         if cycle:
             # There should never be a cycle.  This is just in case.
-            p.text('StateVectorTrialResult(...)')
+            p.text("StateVectorTrialResult(...)")
         else:
             p.text(str(self))
 
     def __repr__(self) -> str:
         return (
-            f'cirq.StateVectorTrialResult(params={self.params!r}, '
-            f'measurements={self.measurements!r}, '
-            f'final_simulator_state={self._final_simulator_state!r})'
+            f"cirq.StateVectorTrialResult(params={self.params!r}, "
+            f"measurements={self.measurements!r}, "
+            f"final_simulator_state={self._final_simulator_state!r})"
         )
 
 

--- a/qsimcirq/qsim_simulator.py
+++ b/qsimcirq/qsim_simulator.py
@@ -43,7 +43,8 @@ class QSimSimulatorState(sim.StateVectorSimulatorState):
         super().__init__(state_vector=state_vector, qubit_map=qubit_map)
 
 
-class QSimSimulatorTrialResult(sim.StateVectorTrialResult):
+@value.value_equality(unhashable=True)
+class QSimSimulatorTrialResult(sim.StateVectorMixin, sim.SimulationTrialResult):
     def __init__(
         self,
         params: study.ParamResolver,
@@ -54,6 +55,47 @@ class QSimSimulatorTrialResult(sim.StateVectorTrialResult):
             params=params,
             measurements=measurements,
             final_simulator_state=final_simulator_state,
+        )
+
+    # The following methods are (temporarily) copied here from
+    # cirq.StateVectorTrialResult due to incompatibility with the
+    # intermediate state simulation support which that class requires.
+    # TODO: remove these methods once inheritance is restored.
+
+    @property
+    def final_state_vector(self):
+        return self._final_simulator_state.state_vector
+
+    def state_vector(self):
+        """Return the state vector at the end of the computation."""
+        return self._final_simulator_state.state_vector.copy()
+
+    def _value_equality_values_(self):
+        measurements = {k: v.tolist() for k, v in sorted(self.measurements.items())}
+        return (self.params, measurements, self._final_simulator_state)
+
+    def __str__(self) -> str:
+        samples = super().__str__()
+        final = self.state_vector()
+        if len([1 for e in final if abs(e) > 0.001]) < 16:
+            state_vector = self.dirac_notation(3)
+        else:
+            state_vector = str(final)
+        return f'measurements: {samples}\noutput vector: {state_vector}'
+
+    def _repr_pretty_(self, p: Any, cycle: bool) -> None:
+        """Text output in Jupyter."""
+        if cycle:
+            # There should never be a cycle.  This is just in case.
+            p.text('StateVectorTrialResult(...)')
+        else:
+            p.text(str(self))
+
+    def __repr__(self) -> str:
+        return (
+            f'cirq.StateVectorTrialResult(params={self.params!r}, '
+            f'measurements={self.measurements!r}, '
+            f'final_simulator_state={self._final_simulator_state!r})'
         )
 
 

